command/lock.go

```diff
@@ -0,0 +1,78 @@
+package command
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform/state"
+)
+
+// LockCommand is a cli.Command implementation that manually locks
+// the state.
+type LockCommand struct {
+	Meta
+}
+
+func (c *LockCommand) Run(args []string) int {
+	args = c.Meta.process(args, false)
+
+	cmdFlags := c.Meta.flagSet("lock")
+	cmdFlags.Usage = func() { c.Ui.Error(c.Help()) }
+	if err := cmdFlags.Parse(args); err != nil {
+		return 1
+	}
+
+	// assume everything is initialized. The user can manually init if this is
+	// required.
+	configPath, err := ModulePath(cmdFlags.Args())
+	if err != nil {
+		c.Ui.Error(err.Error())
+		return 1
+	}
+
+	// Load the backend
+	b, err := c.Backend(&BackendOpts{
+		ConfigPath: configPath,
+	})
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Failed to load backend: %s", err))
+		return 1
+	}
+
+	st, err := b.State()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Failed to load state: %s", err))
+		return 1
+	}
+
+	s, ok := st.(state.Locker)
+	if !ok {
+		c.Ui.Error("Current state does not support locking")
+		return 1
+	}
+
+	if err := s.Lock("lock"); err != nil {
+		c.Ui.Error(fmt.Sprintf("Failed to lock state: %s", err))
+		return 1
+	}
+
+	return 0
+}
+
+func (c *LockCommand) Help() string {
+	helpText := `
+Usage: terraform lock [DIR]
+
+  Manually lock the state for the defined configuration.
+
+  This will not modify your infrastructure. This command obtains a lock on the
+  state for the current configuration. The behavior of this lock is dependent
+  on the backend being used. A lock on a local state file only lasts for the
+  duration of the calling process.
```

Comment:
The symmetry here may work against us by making this feel similar to Subversion locks, explicit database transactions, and other such things that Terraform users are likely to have experience with. This creates a "false cognate" that is likely to lead users astray.

I would define "work" here -- from a user's perspective -- as "create a lock that persists until the user runs `unlock`"... I worry that `terraform lock` command that *can't* guarantee to behave that way would lead people into believing that Terraform has an explicit locking workflow and only later finding that it's been ineffective due to the specific behavior of the backend, possibly after it's too late and data has already been lost.

I think I'd lean towards renaming `unlock` to `force-unlock`, to make it clearer that this is something exceptional that you shouldn't normally need to do, and then eliminating the `lock` command entirely.

If the testing use-case is important then perhaps we can have a `test-lock` command that acquires a lock *and then immediately releases it*, allowing users to exercise the locking behavior (e.g. to validate configuration/credentials) without creating the impression that long-lived locks are an expected workflow.